### ISSUE TITLE
GKE: can not read from google cloud storage in Jupyter notebook

### ISSUE THREAD

**Issue by mmatiaschek** [kubeflow/kubeflow#1249]
the service account that is created for instances seems to have the rights to read from google cloud storage of the project, but i cannot do it in a jupyter notebook:

![image](https://user-images.githubusercontent.com/3010105/43004998-c0e8d4fc-8c31-11e8-8055-bbbfcccc09bb.png)

![image](https://user-images.githubusercontent.com/3010105/43005019-d68b5d84-8c31-11e8-84cc-d06579972fc9.png)

**Comment by jlewi** [comment]
We intentionally locked down the VM instance account to avoid granting broad permissions to all pods running on every VM.

We create GCP service accounts intended for actual work and store the secret as a K8s cluster secret that can be attached to the pods. 

Possible fixes

1. Update the DM configs to grant GCS scope to the VM service account
2. Modify the kube spawner to attach the kubeflow-user service account
3. Download a service account key (e.g. via the gcloud UI) upload it to your pod and then activate it with gcloud.

@ankushagarwal Do we have a way to attach secrets to the Jupyter pods?

**Comment by ankushagarwal** [comment]
Yes, we do inject the [user-gcp-sa](https://github.com/kubeflow/kubeflow/blob/18305be18541b6dcaed1098638e84d603a1b04eb/scripts/gke/deploy.sh#L157) credentials in jupyter notebooks in the latest version of kubeflow. This has the storageAdmin role, so you should be able to access GCS from jupyter notebooks.

**Comment by jlewi** [comment]
@mmatiaschek can you check your pod to see if a secret is attached? If it is you should be able to do something like

```
gcloud auth activate-service-account --key-file=${GOOGLE_APPLICATION_CREDENTIALS}
```

@ankushagarwal I looked at a pod running 0.2.2 and I didn't see a service account attached. Do we need to set cloud or something?

**Comment by jlewi** [comment]
Looks like you need to set cloud
https://github.com/kubeflow/kubeflow/blob/master/kubeflow/core/jupyterhub.libsonnet#L155

And then secret is attached here.
https://github.com/kubeflow/kubeflow/blob/master/kubeflow/core/kubeform_spawner.py#L255

We should do the following

1. The GKE deploy process should do this by default
1. We should document it.

**Comment by mmatiaschek** [comment]
@jlewi i guess it isn't? 
GOOGLE_APPLICATION_CREDENTIALS is not set. `export` does not show me any variable that looks like it is set to a keyfile.
 
```
$ kubectl -n kubeflow exec -it jupyter-accounts-2egoogle-2ecom-3ammatiaschek-40gmail-2ecom -- bash
jovyan@jupyter-accounts-2egoogle-2ecom-3ammatiaschek-40gmail-2ecom:~$ gcloud auth activate-service-account --key-file=${GOOGLE_APPLICATION_CREDENTIALS}
ERROR: (gcloud.auth.activate-service-account) Unable to read file []: [Errno 2] No such file or directory: ''
```

**Comment by jlewi** [comment]
I think you need to set cloud

```
ks param set kubeflow-core cloud gke
ks apply ${ENV} kubeflow-core
```
You'll need to restart your JupyterPod.

**Comment by mmatiaschek** [comment]
unfortunately it is not as easy as
```
ks param set kubeflow-core cloud gke
ks apply default -c kubeflow-core
```

i have just started using ks and still trying to understand how it works... i was doing the ks tutorial when you replied.
i had tried to provision a pd, deleting that i run into another problem while applying the current ks_app:
`ERROR handle object: can't update statefulsets kubeflow.tf-hub: Operation cannot be fulfilled on statefulsets.apps "tf-hub": the object has been modified; please apply your changes to the latest version and try again`
i am currently not aware of any manual changes i did to tf-hub.

**Comment by jlewi** [comment]
This is a bug in ks 0.11 that should be fixed in later version. You just need to retry; e.g. issue ks apply a couple times until it succeeds.

**Comment by jlewi** [comment]
GOOGLE_APPLICATION_CREDENTIALS  should be set if the gcp_secret is set
https://github.com/kubeflow/kubeflow/blob/471de67fc4596fb6727f72332ab555ae10e307eb/kubeflow/core/kubeform_spawner.py#L113

Can you provide the YAML description for your Jupyter pod to see if the secret was actually attached?

```
kubectl get secret user-gcp-sa --namespace hub -o json
```

Will return the base64 encoded value of the key which is probably not what you want.

If the key isn't being mounted automatically into the pod you could do something like the following

You can download a secret key file via the cloud console or via gcloud. See these [instructions](https://cloud.google.com/iam/docs/creating-managing-service-account-keys). 

```
SECRET_FILE=/tmp/somekey.json
EMAIL=${DEPLOYMENT_NAME}-user@${PROJECT}.iam.gserviceaccount.com
gcloud --project=${PROJECT} iam service-accounts keys create ${SECRET_FILE} --iam-account ${EMAIL}
```

You can then upload SECRET_FILE to the Jupyter pod e.g. via the Jupyter UI or using `kubectl cp`.
Once the secret is available inside the pod you could run the `gcloud auth activate-service-account` command.

**Comment by mmatiaschek** [comment]
The workaround you described worked, using `gcloud auth activate-service-account --key-file=...`
Thank you very much!

My kubeform_spawner.py (v0.2.2) doesn't have the change regarding `GOOGLE_APPLICATION_CREDENTIALS`, it only went into master. I guess it will be available in v0.3?
Here's the yaml of the pod:

```
$ kubectl -n hub get po/jupyter-accounts-2egoogle-2ecom-3ammatiaschek-40gmail-2ecom -o yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    hub.jupyter.org/username: accounts.google.com:mmatiaschek@gmail.com
  creationTimestamp: 2018-07-24T09:02:59Z
  labels:
    app: jupyterhub
    component: singleuser-server
    heritage: jupyterhub
  name: jupyter-accounts-2egoogle-2ecom-3ammatiaschek-40gmail-2ecom
  namespace: hub
  resourceVersion: "792779"
  selfLink: /api/v1/namespaces/hub/pods/jupyter-accounts-2egoogle-2ecom-3ammatiaschek-40gmail-2ecom
  uid: 5cadade4-8f20-11e8-9a42-42010a840fce
spec:
  automountServiceAccountToken: false
  containers:
  - args:
    - start-singleuser.sh
    - --ip="0.0.0.0"
    - --port=8888
    - --allow-root
    env:
    - name: JUPYTERHUB_API_TOKEN
      value: bdc2xxxxxxxxxxxxxxx9ed64be
    - name: JPY_API_TOKEN
      value: bdc23933xxxxxxxxxxxxxxxd64be
    - name: JUPYTERHUB_CLIENT_ID
      value: jupyterhub-user-accounts.google.com%3Ammatiaschek%40gmail.com
    - name: JUPYTERHUB_HOST
    - name: JUPYTERHUB_OAUTH_CALLBACK_URL
      value: /user/accounts.google.com%3Ammatiaschek@gmail.com/oauth_callback
    - name: JUPYTERHUB_USER
      value: accounts.google.com:mmatiaschek@gmail.com
    - name: JUPYTERHUB_API_URL
      value: http://tf-hub-0:8081/hub/api
    - name: JUPYTERHUB_BASE_URL
      value: /
    - name: JUPYTERHUB_SERVICE_PREFIX
      value: /user/accounts.google.com%3Ammatiaschek@gmail.com/
    - name: MEM_GUARANTEE
      value: 8Gi
    - name: CPU_GUARANTEE
      value: "2.0"
    image: gcr.io/kubeflow-images-public/tensorflow-1.8.0-notebook-gpu:v0.2.1
    imagePullPolicy: IfNotPresent
    lifecycle: {}
    name: notebook
    ports:
    - containerPort: 8888
      name: notebook-port
      protocol: TCP
    resources:
      limits:
        nvidia.com/gpu: "1"
      requests:
        cpu: "2"
        memory: 8Gi
        nvidia.com/gpu: "1"
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /home/jovyan
      name: volume-accounts-2egoogle-2ecom-3ammatiaschek-40gmail-2ecom
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: no-api-access-please
      readOnly: true
    workingDir: /home/jovyan
  dnsPolicy: ClusterFirst
  nodeName: gke-hub-gpu-pool-67d709c9-h1cf
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext:
    fsGroup: 100
    runAsUser: 1000
  serviceAccount: default
  serviceAccountName: default
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  - effect: NoSchedule
    key: nvidia.com/gpu
    operator: Exists
  volumes:
  - name: volume-accounts-2egoogle-2ecom-3ammatiaschek-40gmail-2ecom
    persistentVolumeClaim:
      claimName: claim-accounts-2egoogle-2ecom-3ammatiaschek-40gmail-2ecom
  - emptyDir: {}
    name: no-api-access-please
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: 2018-07-24T09:02:59Z
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: 2018-07-24T09:03:09Z
    status: "True"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: 2018-07-24T09:02:59Z
    status: "True"
    type: PodScheduled
  containerStatuses:
  - containerID: docker://c41540fc286a027990ea5940e14b4057311381604993187bc8b16eb3bbd4856f
    image: gcr.io/kubeflow-images-public/tensorflow-1.8.0-notebook-gpu:v0.2.1
    imageID: docker-pullable://gcr.io/kubeflow-images-public/tensorflow-1.8.0-notebook-gpu@sha256:63b30b10b0b597a4d52701c8bb2cfa40826ce39f9b33d66a3648c30e2e0e28d3
    lastState: {}
    name: notebook
    ready: true
    restartCount: 0
    state:
      running:
        startedAt: 2018-07-24T09:03:09Z
  hostIP: 10.132.0.4
  phase: Running
  podIP: 10.60.2.4
  qosClass: Burstable
  startTime: 2018-07-24T09:02:59Z
```

`/var/run/secrets/kubernetes.io/serviceaccount` is empty

**Comment by jlewi** [comment]
@mmatiaschek Thank you so much for bearing with us. Yes the fix will be in the next release 0.3.

**Comment by jlewi** [comment]
This should be fixed in 0.3.
